### ISSUE TITLE
Low and High bound of a Box action_space are not respected

### ISSUE THREAD

**Issue by mecotiere** [ray-project/ray#2965]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 18.04.1
- **Ray installed from (source or binary)**: pip
- **Ray version**: 0.5.2
- **Python version**: 3.6
- **Exact command to reproduce**: Any Box action_space eg : 
self.action_space = Box(low=-2.0,
                                high=0.0,
                                shape=(1,),
                                dtype=np.float32
                                )

### Describe the problem

Code example :

    def step(self, action):
        print(action)

The printed action doesn't respect the low & high given when the action_space was initialized.

**Comment by ericl** [comment]
Hey @mecotiere , if your environment requires you can force the action range with `"model": {"squash_to_range": true}`. This will squash the outputs of the policy into the specified range with a sigmoid function.

See also: https://github.com/ray-project/ray/issues/1862

**Comment by ericl** [comment]
Oh, good catch. That's a bug in the PPO option propagation, fix here: https://github.com/ray-project/ray/pull/2974

**Comment by fruffy** [comment]
I am digging this issue out again since our problem is quite related. I see similar behaviour with DDPG. We set our action space from 0 to 1. PPO and PG respect the bounds, but DDPG does not. It often turns negative which translates to an invalid action. How could we fix this?
How many algorithms could be affected by this?

**Comment by joneswong** [comment]
Oh, I see. For Gaussian, we use `tf.clip_by_value`. For OU, even though `deterministic_actions` is squashed by sigmoid function, the posed OU noise may make it out of valid range. There are some environments that have not strictly obey gym interface, say that miss the action clipping step. It is necessary to also call `clip_by_value` op in OU case, imo. 
Besides, the saturation problem of tanh/sigmoid is also studied in some recent papers, like https://arxiv.org/abs/1511.04143    Do you think such a configurable feature is necessary?

**Comment by ericl** [comment]
@joneswong makes sense. I guess unlike PPO, there probably isn't a reason to avoid clipping since this part isn't differentiable anyways. So we don't need to make it configurable.

https://github.com/ray-project/ray/pull/3386/files should fix it

**Comment by ericl** [comment]
Hm actually maybe we should clip_by_value for PPO as well.

**Comment by fruffy** [comment]
Great! Thanks as always guys for reacting so quickly.

**Comment by ericl** [comment]
https://github.com/ray-project/ray/pull/3426 remove squash_to_range which is buggy in favor of a plain action_clip
